library/std/src/panic.rs

```diff
@@ -202,5 +203,99 @@ pub fn always_abort() {
+    /// Backtraces aren't supported on this platform.
+    Unsupported,
+}
+
+impl BacktraceStyle {
+    pub(crate) fn full() -> Self {
+        if cfg!(feature = "backtrace") { BacktraceStyle::Full } else { BacktraceStyle::Disabled }
+    }
+
+    fn as_usize(self) -> usize {
+        match self {
+            BacktraceStyle::Short => 1,
+            BacktraceStyle::Full => 2,
+            BacktraceStyle::Disabled => 3,
+            BacktraceStyle::Unsupported => 4,
+        }
+    }
+}
+
+// Tracks whether we should/can capture a backtrace, and how we should display
+// that backtrace.
+//
+// Internally stores equivalent of an Option<BacktraceStyle>.
+static SHOULD_CAPTURE: AtomicUsize = AtomicUsize::new(0);
+
+/// Configure whether the default panic hook will capture and display a
+/// backtrace.
+///
+/// This is also configurable via the environment variable `RUST_BACKTRACE`, but
+/// usage of this function should be preferred from within Rust code.
+#[unstable(feature = "panic_backtrace_config", issue = "93346")]
+pub fn set_backtrace_style(style: BacktraceStyle) {
+    SHOULD_CAPTURE.store(style.as_usize(), Ordering::Release);
+}
+
+/// Checks whether the standard library's panic hook will capture and print a
+/// backtrace.
+///
+/// Note that `set_backtrace_style` may not change the result of this function,
+/// for example when backtraces are not supported on the current platform.
+///
+/// This is intended primarily to allow customized panic hooks to not need their
+/// own configuration for backtrace printing, instead letting them rely on the
+/// same configuration provided by the standard library.
+#[unstable(feature = "panic_backtrace_config", issue = "93346")]
+pub fn get_backtrace_style() -> BacktraceStyle {
+    if !cfg!(feature = "backtrace") {
+        // If the `backtrace` feature of this crate isn't enabled quickly return
+        // `Disabled` so this can be constant propagated all over the place to
+        // optimize away callers.
+        return BacktraceStyle::Unsupported;
+    }
+    match SHOULD_CAPTURE.load(Ordering::Acquire) {
+        0 => {}
+        1 => return BacktraceStyle::Short,
+        2 => return BacktraceStyle::Full,
+        3 => return BacktraceStyle::Disabled,
+        _ => unreachable!(),
+    }
+
+    // Setting environment variables for Fuchsia components isn't a standard
+    // or easily supported workflow. For now, always display backtraces.
+    let format = if cfg!(target_os = "fuchsia") {
+        BacktraceStyle::Full
+    } else {
+        crate::env::var_os("RUST_BACKTRACE")
+            .map(|x| {
+                if &x == "0" {
+                    BacktraceStyle::Disabled
+                } else if &x == "full" {
+                    BacktraceStyle::Full
+                } else {
+                    BacktraceStyle::Short
+                }
+            })
+            .unwrap_or(BacktraceStyle::Disabled)
+    };
+    SHOULD_CAPTURE.store(format.as_usize(), Ordering::Release);
```

Comment:
I think this should use `set_backtrace_style` so that any validation/mapping that that function does is also done here.